src/compiler/parser.ts

```diff
@@ -1001,10 +1001,10 @@ module ts {
             switch (parsingContext) {
                 case ParsingContext.SourceElements:
                 case ParsingContext.ModuleElements:
-                    return isSourceElement(inErrorRecovery);
+                    return !(token === SyntaxKind.SemicolonToken && inErrorRecovery) && isModuleElement();

```

Comment:
Helper or not, at the very least, there should be a comment above the function explaining what `inErrorRecovery` implies about the behavior of `isListElement`.